// ng-remote-autocomplete/bundles/ng-remote-autocomplete.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs/Subject'), require('@angular/forms'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ng-remote-autocomplete', ['exports', '@angular/core', 'rxjs/Subject', '@angular/forms', '@angular/common'], factory) :
    (factory((global['ng-remote-autocomplete'] = {}),global.ng.core,global.rxjs.Subject,global.ng.forms,global.ng.common));
}(this, (function (exports,core,Subject,forms,common) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var DEFAULT_ACTIVE_INDEX = -1;
    /** @type {?} */
    var DEFAULT_MIN_SEARCH_LENGTH = 1;
    /** @type {?} */
    var DEFAULT_MAX_CHARS = 2147483647;
    /** @type {?} */
    var DEFAULT_PAUSE = 100;
    /** @type {?} */
    var NOT_FOUND_TEXT = "No results found";
    /** @type {?} */
    var SEARCHING_TEXT = "Searching...";
    /** @type {?} */
    var SEARCH_WRAPPER_CLASS = "autocomplete-search-wrapper";
    /** @type {?} */
    var OPTION_VALUE_CLASS = "autocomplete-option-value";
    /** @type {?} */
    var OPTION_DESCRIPTION_CLASS = "autocomplete-option-description";
    /** @type {?} */
    var OPTION_CLASS = "autocomplete-option";
    /** @type {?} */
    var OPTION_LIST_WRAPPER_CLASS = "option-list-wrapper";
    /** @type {?} */
    var LOADING_CLASS = "autocomplete-loading";
    /** @type {?} */
    var RESULTS_NOT_FOUND_CLASS = "autocomplete-not-found";

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {number} */
    var SearchStateType = {
        Untracked: 0,
        Loading: 1,
        Finished: 2,
    };
    SearchStateType[SearchStateType.Untracked] = 'Untracked';
    SearchStateType[SearchStateType.Loading] = 'Loading';
    SearchStateType[SearchStateType.Finished] = 'Finished';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var OptionTemplateDirective = (function () {
        function OptionTemplateDirective(template) {
            this.template = template;
        }
        OptionTemplateDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[remote-autocomplete-option-tmpl]'
                    },] },
        ];
        OptionTemplateDirective.ctorParameters = function () {
            return [
                { type: core.TemplateRef }
            ];
        };
        return OptionTemplateDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var AutocompleteService = (function () {
        function AutocompleteService() {
            this.changedHighlightItemSubject = new Subject.Subject();
            this.selectItemSubject = new Subject.Subject();
            this._isOpen = false;
        }
        Object.defineProperty(AutocompleteService.prototype, "isOpen", {
            get: /**
             * @return {?}
             */ function () {
                return this._isOpen;
            },
            set: /**
             * @param {?} value
             * @return {?}
             */ function (value) {
                this._isOpen = value;
            },
            enumerable: true,
            configurable: true
        });
        AutocompleteService.decorators = [
            { type: core.Injectable },
        ];
        return AutocompleteService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ItemListService = (function () {
        function ItemListService(autocompleteService) {
            this.autocompleteService = autocompleteService;
            this._activeIndex = DEFAULT_ACTIVE_INDEX;
            this.items = [];
        }
        Object.defineProperty(ItemListService.prototype, "activeIndex", {
            get: /**
             * @return {?}
             */ function () {
                return this._activeIndex;
            },
            set: /**
             * @param {?} index
             * @return {?}
             */ function (index) {
                if (this.autocompleteService.isOpen) {
                    if (index >= this.items.length) {
                        this.restoreIntialActiveIndex();
                        return;
                    }
                    if (index < DEFAULT_ACTIVE_INDEX) {
                        this._activeIndex = this.items.length - 1;
                        this.autocompleteService.changedHighlightItemSubject.next(this._activeIndex);
                        return;
                    }
                    this._activeIndex = index;
                    this.autocompleteService.changedHighlightItemSubject.next(this._activeIndex);
                }
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        ItemListService.prototype.restoreIntialActiveIndex = /**
         * @return {?}
         */
            function () {
                this._activeIndex = DEFAULT_ACTIVE_INDEX;
                this.autocompleteService.changedHighlightItemSubject.next(this._activeIndex);
            };
        ItemListService.decorators = [
            { type: core.Injectable },
        ];
        ItemListService.ctorParameters = function () {
            return [
                { type: AutocompleteService }
            ];
        };
        return ItemListService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var RemoteAutocompleteComponent = (function () {
        function RemoteAutocompleteComponent(itemListService, autocompleteService) {
            this.itemListService = itemListService;
            this.autocompleteService = autocompleteService;
            this.stopEnterPropagationOnSelect = false;
            this.needToShowNotFoundText = true;
            this.needToShowSearchingText = true;
            this.type = new core.EventEmitter();
            this.highlighted = new core.EventEmitter();
            this.selected = new core.EventEmitter();
            this.blur = new core.EventEmitter();
            this.focusEvent = new core.EventEmitter();
            this.searchStates = SearchStateType;
            this.searchState = SearchStateType.Untracked;
            this.SEARCH_WRAPPER_CLASS = SEARCH_WRAPPER_CLASS;
            this.OPTION_LIST_WRAPPER_CLASS = OPTION_LIST_WRAPPER_CLASS;
            this.OPTION_CLASS = OPTION_CLASS;
            this.OPTION_VALUE_CLASS = OPTION_VALUE_CLASS;
            this.OPTION_DESCRIPTION_CLASS = OPTION_DESCRIPTION_CLASS;
            this.LOADING_CLASS = LOADING_CLASS;
            this.RESULTS_NOT_FOUND_CLASS = RESULTS_NOT_FOUND_CLASS;
            this.propagateChange = (function (_) { });
            this.propagateTouched = (function (_) { });
        }
        /**
         * @return {?}
         */
        RemoteAutocompleteComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                var _this = this;
                this.validateService();
                this.disabled = this.getDefaultIfNullOrUndefined(this.disabled, false);
                this.minChars = this.getDefaultIfNullOrUndefined(this.minChars, DEFAULT_MIN_SEARCH_LENGTH);
                this.maxChars = this.maxChars || DEFAULT_MAX_CHARS;
                this.pause = this.pause || DEFAULT_PAUSE;
                this.notFoundText = this.notFoundText || NOT_FOUND_TEXT;
                this.searchingText = this.searchingText || SEARCHING_TEXT;
                this.changedHighlightSubscription = this.autocompleteService.changedHighlightItemSubject.subscribe((function (index) {
                    _this.onActiveIndexChanged(index);
                }));
                this.selectSubscription = this.autocompleteService.selectItemSubject.subscribe((function (index) {
                    /** @type {?} */
                    var result = _this.searchResult[index];
                    _this.propagateChange(result.value);
                    _this.selected.emit(result);
                    _this.close();
                }));
            };
        /**
         * @return {?}
         */
        RemoteAutocompleteComponent.prototype.ngOnDestroy = /**
         * @return {?}
         */
            function () {
                this.changedHighlightSubscription.unsubscribe();
                this.selectSubscription.unsubscribe();
            };
        /**
         * @param {?} value
         * @return {?}
         */
        RemoteAutocompleteComponent.prototype.writeValue = /**
         * @param {?} value
         * @return {?}
         */
            function (value) {
                if (value !== undefined) {
                    this.searchValue = value;
                }
            };
        /**
         * @param {?} fn
         * @return {?}
         */
        RemoteAutocompleteComponent.prototype.registerOnChange = /**
         * @param {?} fn
         * @return {?}
         */
            function (fn) {
                this.propagateChange = fn;
            };
        /**
         * @param {?} fn
         * @return {?}
         */
        RemoteAutocompleteComponent.prototype.registerOnTouched = /**
         * @param {?} fn
         * @return {?}
         */
            function (fn) {
                this.propagateTouched = fn;
            };
        /**
         * @param {?} isDisabled
         * @return {?}
         */
        RemoteAutocompleteComponent.prototype.setDisabledState = /**
         * @param {?} isDisabled
         * @return {?}
         */
            function (isDisabled) {
                this.disabled = isDisabled;
            };
        /**
         * @return {?}
         */
        RemoteAutocompleteComponent.prototype.onType = /**
         * @return {?}
         */
            function () {
                this.propagateChange(this.searchValue);
                this.originalSearchValue = this.searchValue;
                this.type.emit();
                if (this.validsearchTerm()) {
                    this.searchState = this.searchStates.Loading;
                }
                else {
                    this.close();
                }
                this.resetSearchTimeout();
            };
        /**
         * @param {?} event
         * @return {?}
         */
        RemoteAutocompleteComponent.prototype.onInputBlur = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                if (event.relatedTarget) {
                    /** @type {?} */
                    var related = ((event.relatedTarget));
                    if (related.classList.contains(this.OPTION_CLASS)) {
                        return;
                    }
                }
                this.propagateTouched(this.searchValue);
                this.close();
                this.blur.emit(event);
            };
        /**
         * @param {?} event
         * @return {?}
         */
        RemoteAutocompleteComponent.prototype.onInputFocus = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                this.focusEvent.emit(event);
            };
        Object.defineProperty(RemoteAutocompleteComponent.prototype, "inputClass", {
            get: /**
             * @return {?}
             */ function () {
                /** @type {?} */
                var classes = '';
                if (this.inputClasses) {
                    classes = this.inputClasses.join(' ');
                }
                return classes;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        RemoteAutocompleteComponent.prototype.focus = /**
         * @return {?}
         */
            function () {
                (((this.inputField.nativeElement))).focus();
            };
        /**
         * @return {?}
         */
        RemoteAutocompleteComponent.prototype.open = /**
         * @return {?}
         */
            function () {
                this.onType();
            };
        /**
         * @return {?}
         */
        RemoteAutocompleteComponent.prototype.close = /**
         * @return {?}
         */
            function () {
                this.autocompleteService.isOpen = false;
                this.searchState = this.searchStates.Untracked;
            };
        /**
         * @private
         * @return {?}
         */
        RemoteAutocompleteComponent.prototype.validateService = /**
         * @private
         * @return {?}
         */
            function () {
                if (!this.service) {
                    throw new Error("AutocompleteService must be implemented");
                }
                if (this.isParameterizedSearch(this.service) && (this.searchParameters === undefined || this.searchParameters === null)) {
                    throw new Error("Please provide additionalSearchParams or just provide AutocompleteSourceService instaead of ParameterizedAutocompleteSourceService");
                }
            };
        /**
         * @private
         * @param {?} activeIndex
         * @return {?}
         */
        RemoteAutocompleteComponent.prototype.onActiveIndexChanged = /**
         * @private
         * @param {?} activeIndex
         * @return {?}
         */
            function (activeIndex) {
                /** @type {?} */
                var highlightedItem;
                if (activeIndex === DEFAULT_ACTIVE_INDEX) {
                    highlightedItem = null;
                    this.searchValue = this.originalSearchValue;
                }
                else {
                    highlightedItem = this.searchResult[activeIndex];
                    this.searchValue = highlightedItem.value;
                }
                this.propagateChange(this.searchValue);
                this.highlighted.emit(highlightedItem);
            };
        /**
         * @private
         * @return {?}
         */
        RemoteAutocompleteComponent.prototype.resetSearchTimeout = /**
         * @private
         * @return {?}
         */
            function () {
                this.clearSearchTimeout();
                this.setSearchTimeout();
            };
        /**
         * @private
         * @return {?}
         */
        RemoteAutocompleteComponent.prototype.setSearchTimeout = /**
         * @private
         * @return {?}
         */
            function () {
                this.typingTimeout = setTimeout(this.search.bind(this), this.pause);
            };
        /**
         * @private
         * @return {?}
         */
        RemoteAutocompleteComponent.prototype.clearSearchTimeout = /**
         * @private
         * @return {?}
         */
            function () {
                clearTimeout(this.typingTimeout);
            };
        /**
         * @private
         * @return {?}
         */
        RemoteAutocompleteComponent.prototype.search = /**
         * @private
         * @return {?}
         */
            function () {
                var _this = this;
                this.itemListService.restoreIntialActiveIndex();
                if (this.validsearchTerm()) {
                    this.performSearch().first().subscribe((function (results) {
                        if (_this.searchState !== _this.searchStates.Untracked) {
                            _this.autocompleteService.isOpen = results.length > 0;
                            _this.itemListService.items = results;
                            _this.searchResult = results;
                            _this.searchState = _this.searchStates.Finished;
                        }
                    }));
                }
            };
        /**
         * @private
         * @return {?}
         */
        RemoteAutocompleteComponent.prototype.performSearch = /**
         * @private
         * @return {?}
         */
            function () {
                if (this.isParameterizedSearch(this.service)) {
                    return (((this.service))).getWithParams(this.searchValue, this.searchParameters);
                }
                else {
                    return (((this.service))).get(this.searchValue);
                }
            };
        /**
         * @private
         * @return {?}
         */
        RemoteAutocompleteComponent.prototype.validsearchTerm = /**
         * @private
         * @return {?}
         */
            function () {
                return (this.searchValue || '').length >= this.minChars;
            };
        /**
         * @private
         * @param {?} service
         * @return {?}
         */
        RemoteAutocompleteComponent.prototype.isParameterizedSearch = /**
         * @private
         * @param {?} service
         * @return {?}
         */
            function (service) {
                return (((service))).getWithParams !== undefined;
            };
        /**
         * @private
         * @template T
         * @param {?} value
         * @param {?} defaultValue
         * @return {?}
         */
        RemoteAutocompleteComponent.prototype.getDefaultIfNullOrUndefined = /**
         * @private
         * @template T
         * @param {?} value
         * @param {?} defaultValue
         * @return {?}
         */
            function (value, defaultValue) {
                return value === undefined || value === null
                    ? defaultValue
                    : value;
            };
        RemoteAutocompleteComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'remote-autocomplete',
                        template: "<div keyboard-navigation [stopEnterPropagation]=\"stopEnterPropagationOnSelect\" class=\"remote-autocomplete\">\n    <input [(ngModel)]=\"searchValue\"\n           [attr.id]=\"inputId\"\n           [attr.name]=\"inputName\"\n           [attr.class]=\"inputClass\"\n           [disabled]=\"disabled\"\n           [attr.placeholder]=\"placeholder\"\n           [attr.autocomplete]=\"autocompleteAttr\"\n           [attr.tabindex]=\"tabindex\"\n           maxlength=\"{{maxChars}}\"\n           (input)=\"onType()\"\n           (blur)=\"onInputBlur($event)\"\n           (focus)=\"onInputFocus($event)\"\n           #input />\n    <div *ngIf=\"searchState === searchStates.Loading && needToShowSearchingText || \n                searchState === searchStates.Finished && (searchResult?.length > 0 || needToShowNotFoundText)\" [class]=\"SEARCH_WRAPPER_CLASS\">\n        <div *ngIf=\"searchState === searchStates.Loading\">\n            <span [class]=\"LOADING_CLASS\">{{searchingText}}</span>\n        </div>\n\n        <div *ngIf=\"searchState === searchStates.Finished && searchResult.length === 0\">\n            <span [class]=\"RESULTS_NOT_FOUND_CLASS\">{{notFoundText}}</span>\n        </div>\n\n        <div *ngIf=\"searchState === searchStates.Finished && searchResult?.length > 0\" [class]=\"OPTION_LIST_WRAPPER_CLASS\">\n            <div *ngFor=\"let item of searchResult; index as i\"\n                 remote-autocomplete-option\n                 [itemIndex]=\"i\"\n                 [class]=\"OPTION_CLASS\"\n                 [ngClass]=\"{'autocomplete-option-active': i === itemListService.activeIndex}\"\n                 tabindex=\"-1\">\n                <ng-template #defaultOptionTemplate>\n                    <span [class]=\"OPTION_VALUE_CLASS\">{{item.value}}</span>\n                    <span *ngIf=\"item.description\" [class]=\"OPTION_DESCRIPTION_CLASS\">{{item.description}}</span>\n                </ng-template>\n\n                <ng-template [ngTemplateOutlet]=\"optionTemplate || defaultOptionTemplate\" [ngTemplateOutletContext]=\"{ $implicit: item }\">\n                </ng-template>\n            </div>\n        </div>\n    </div>\n</div>",
                        styles: [".autocomplete-search-wrapper{border:1px solid #ccc;max-height:200px;overflow:auto;max-width:200px}.autocomplete-option-active{background-color:#d3d3d3}"],
                        providers: [
                            AutocompleteService,
                            ItemListService,
                            {
                                provide: forms.NG_VALUE_ACCESSOR,
                                useExisting: core.forwardRef((function () { return RemoteAutocompleteComponent; })),
                                multi: true
                            }
                        ]
                    },] },
        ];
        RemoteAutocompleteComponent.ctorParameters = function () {
            return [
                { type: ItemListService },
                { type: AutocompleteService }
            ];
        };
        RemoteAutocompleteComponent.propDecorators = {
            service: [{ type: core.Input, args: ['service',] }],
            minChars: [{ type: core.Input, args: ['minSearchLength',] }],
            maxChars: [{ type: core.Input, args: ['maxChars',] }],
            pause: [{ type: core.Input, args: ['pause',] }],
            inputId: [{ type: core.Input, args: ['inputId',] }],
            inputName: [{ type: core.Input, args: ['inputName',] }],
            inputClasses: [{ type: core.Input, args: ['inputClass',] }],
            disabled: [{ type: core.Input, args: ['disabled',] }],
            notFoundText: [{ type: core.Input, args: ['notFoundText',] }],
            searchingText: [{ type: core.Input, args: ['searchingText',] }],
            placeholder: [{ type: core.Input, args: ['placeholder',] }],
            searchParameters: [{ type: core.Input, args: ['searchParameters',] }],
            autocompleteAttr: [{ type: core.Input, args: ['autocompleteAttr',] }],
            tabindex: [{ type: core.Input, args: ['tabindex',] }],
            stopEnterPropagationOnSelect: [{ type: core.Input, args: ['stopEnterPropagationOnSelect',] }],
            needToShowNotFoundText: [{ type: core.Input, args: ['needToShowNotFoundText',] }],
            needToShowSearchingText: [{ type: core.Input, args: ['needToShowSearchingText',] }],
            type: [{ type: core.Output, args: ['type',] }],
            highlighted: [{ type: core.Output, args: ['highlighted',] }],
            selected: [{ type: core.Output, args: ['selected',] }],
            blur: [{ type: core.Output, args: ['blur',] }],
            focusEvent: [{ type: core.Output, args: ['focus',] }],
            optionTemplate: [{ type: core.ContentChild, args: [OptionTemplateDirective, { read: core.TemplateRef },] }],
            inputField: [{ type: core.ViewChild, args: ['input',] }]
        };
        return RemoteAutocompleteComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var OptionDirective = (function () {
        function OptionDirective(autocompleteService, itemListService) {
            this.autocompleteService = autocompleteService;
            this.itemListService = itemListService;
        }
        /**
         * @param {?} event
         * @return {?}
         */
        OptionDirective.prototype.onMouseEnter = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                this.itemListService.activeIndex = this.index;
            };
        /**
         * @param {?} event
         * @return {?}
         */
        OptionDirective.prototype.onMouseLeave = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                this.itemListService.restoreIntialActiveIndex();
            };
        /**
         * @param {?} event
         * @return {?}
         */
        OptionDirective.prototype.onClick = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                this.autocompleteService.selectItemSubject.next(this.index);
            };
        OptionDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[remote-autocomplete-option]',
                        host: {
                            '(mouseenter)': 'onMouseEnter($event)',
                            '(mouseleave)': 'onMouseLeave($event)',
                            '(click)': 'onClick($event)'
                        }
                    },] },
        ];
        OptionDirective.ctorParameters = function () {
            return [
                { type: AutocompleteService },
                { type: ItemListService }
            ];
        };
        OptionDirective.propDecorators = {
            index: [{ type: core.Input, args: ['itemIndex',] }]
        };
        return OptionDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {number} */
    var KeyCode = {
        Backspace: 8,
        Tab: 9,
        Enter: 13,
        Shift: 16,
        Ctrl: 17,
        Alt: 18,
        PauseBreak: 19,
        CapsLock: 20,
        Escape: 27,
        Space: 32,
        PageUp: 33,
        PageDown: 34,
        End: 35,
        Home: 36,
        LeftArrow: 37,
        UpArrow: 38,
        RightArrow: 39,
        DownArrow: 40,
        Insert: 45,
        Delete: 46,
        Zero: 48,
        ClosedParen: 48,
        One: 49,
        ExclamationMark: 49,
        Two: 50,
        AtSign: 50,
        Three: 51,
        PoundSign: 51,
        Hash: 51,
        Four: 52,
        DollarSign: 52,
        Five: 53,
        PercentSign: 53,
        Six: 54,
        Caret: 54,
        Hat: 54,
        Seven: 55,
        Ampersand: 55,
        Eight: 56,
        Star: 56,
        Asterik: 56,
        Nine: 57,
        OpenParen: 57,
        A: 65,
        B: 66,
        C: 67,
        D: 68,
        E: 69,
        F: 70,
        G: 71,
        H: 72,
        I: 73,
        J: 74,
        K: 75,
        L: 76,
        M: 77,
        N: 78,
        O: 79,
        P: 80,
        Q: 81,
        R: 82,
        S: 83,
        T: 84,
        U: 85,
        V: 86,
        W: 87,
        X: 88,
        Y: 89,
        Z: 90,
        LeftWindowKey: 91,
        RightWindowKey: 92,
        SelectKey: 93,
        Numpad0: 96,
        Numpad1: 97,
        Numpad2: 98,
        Numpad3: 99,
        Numpad4: 100,
        Numpad5: 101,
        Numpad6: 102,
        Numpad7: 103,
        Numpad8: 104,
        Numpad9: 105,
        Multiply: 106,
        Add: 107,
        Subtract: 109,
        DecimalPoint: 110,
        Divide: 111,
        F1: 112,
        F2: 113,
        F3: 114,
        F4: 115,
        F5: 116,
        F6: 117,
        F7: 118,
        F8: 119,
        F9: 120,
        F10: 121,
        F11: 122,
        F12: 123,
        NumLock: 144,
        ScrollLock: 145,
        SemiColon: 186,
        Equals: 187,
        Comma: 188,
        Dash: 189,
        Period: 190,
        UnderScore: 189,
        PlusSign: 187,
        ForwardSlash: 191,
        Tilde: 192,
        GraveAccent: 192,
        OpenBracket: 219,
        ClosedBracket: 221,
        Quote: 222,
    };
    KeyCode[KeyCode.Backspace] = 'Backspace';
    KeyCode[KeyCode.Tab] = 'Tab';
    KeyCode[KeyCode.Enter] = 'Enter';
    KeyCode[KeyCode.Shift] = 'Shift';
    KeyCode[KeyCode.Ctrl] = 'Ctrl';
    KeyCode[KeyCode.Alt] = 'Alt';
    KeyCode[KeyCode.PauseBreak] = 'PauseBreak';
    KeyCode[KeyCode.CapsLock] = 'CapsLock';
    KeyCode[KeyCode.Escape] = 'Escape';
    KeyCode[KeyCode.Space] = 'Space';
    KeyCode[KeyCode.PageUp] = 'PageUp';
    KeyCode[KeyCode.PageDown] = 'PageDown';
    KeyCode[KeyCode.End] = 'End';
    KeyCode[KeyCode.Home] = 'Home';
    KeyCode[KeyCode.LeftArrow] = 'LeftArrow';
    KeyCode[KeyCode.UpArrow] = 'UpArrow';
    KeyCode[KeyCode.RightArrow] = 'RightArrow';
    KeyCode[KeyCode.DownArrow] = 'DownArrow';
    KeyCode[KeyCode.Insert] = 'Insert';
    KeyCode[KeyCode.Delete] = 'Delete';
    KeyCode[KeyCode.Zero] = 'Zero';
    KeyCode[KeyCode.ClosedParen] = 'ClosedParen';
    KeyCode[KeyCode.One] = 'One';
    KeyCode[KeyCode.ExclamationMark] = 'ExclamationMark';
    KeyCode[KeyCode.Two] = 'Two';
    KeyCode[KeyCode.AtSign] = 'AtSign';
    KeyCode[KeyCode.Three] = 'Three';
    KeyCode[KeyCode.PoundSign] = 'PoundSign';
    KeyCode[KeyCode.Hash] = 'Hash';
    KeyCode[KeyCode.Four] = 'Four';
    KeyCode[KeyCode.DollarSign] = 'DollarSign';
    KeyCode[KeyCode.Five] = 'Five';
    KeyCode[KeyCode.PercentSign] = 'PercentSign';
    KeyCode[KeyCode.Six] = 'Six';
    KeyCode[KeyCode.Caret] = 'Caret';
    KeyCode[KeyCode.Hat] = 'Hat';
    KeyCode[KeyCode.Seven] = 'Seven';
    KeyCode[KeyCode.Ampersand] = 'Ampersand';
    KeyCode[KeyCode.Eight] = 'Eight';
    KeyCode[KeyCode.Star] = 'Star';
    KeyCode[KeyCode.Asterik] = 'Asterik';
    KeyCode[KeyCode.Nine] = 'Nine';
    KeyCode[KeyCode.OpenParen] = 'OpenParen';
    KeyCode[KeyCode.A] = 'A';
    KeyCode[KeyCode.B] = 'B';
    KeyCode[KeyCode.C] = 'C';
    KeyCode[KeyCode.D] = 'D';
    KeyCode[KeyCode.E] = 'E';
    KeyCode[KeyCode.F] = 'F';
    KeyCode[KeyCode.G] = 'G';
    KeyCode[KeyCode.H] = 'H';
    KeyCode[KeyCode.I] = 'I';
    KeyCode[KeyCode.J] = 'J';
    KeyCode[KeyCode.K] = 'K';
    KeyCode[KeyCode.L] = 'L';
    KeyCode[KeyCode.M] = 'M';
    KeyCode[KeyCode.N] = 'N';
    KeyCode[KeyCode.O] = 'O';
    KeyCode[KeyCode.P] = 'P';
    KeyCode[KeyCode.Q] = 'Q';
    KeyCode[KeyCode.R] = 'R';
    KeyCode[KeyCode.S] = 'S';
    KeyCode[KeyCode.T] = 'T';
    KeyCode[KeyCode.U] = 'U';
    KeyCode[KeyCode.V] = 'V';
    KeyCode[KeyCode.W] = 'W';
    KeyCode[KeyCode.X] = 'X';
    KeyCode[KeyCode.Y] = 'Y';
    KeyCode[KeyCode.Z] = 'Z';
    KeyCode[KeyCode.LeftWindowKey] = 'LeftWindowKey';
    KeyCode[KeyCode.RightWindowKey] = 'RightWindowKey';
    KeyCode[KeyCode.SelectKey] = 'SelectKey';
    KeyCode[KeyCode.Numpad0] = 'Numpad0';
    KeyCode[KeyCode.Numpad1] = 'Numpad1';
    KeyCode[KeyCode.Numpad2] = 'Numpad2';
    KeyCode[KeyCode.Numpad3] = 'Numpad3';
    KeyCode[KeyCode.Numpad4] = 'Numpad4';
    KeyCode[KeyCode.Numpad5] = 'Numpad5';
    KeyCode[KeyCode.Numpad6] = 'Numpad6';
    KeyCode[KeyCode.Numpad7] = 'Numpad7';
    KeyCode[KeyCode.Numpad8] = 'Numpad8';
    KeyCode[KeyCode.Numpad9] = 'Numpad9';
    KeyCode[KeyCode.Multiply] = 'Multiply';
    KeyCode[KeyCode.Add] = 'Add';
    KeyCode[KeyCode.Subtract] = 'Subtract';
    KeyCode[KeyCode.DecimalPoint] = 'DecimalPoint';
    KeyCode[KeyCode.Divide] = 'Divide';
    KeyCode[KeyCode.F1] = 'F1';
    KeyCode[KeyCode.F2] = 'F2';
    KeyCode[KeyCode.F3] = 'F3';
    KeyCode[KeyCode.F4] = 'F4';
    KeyCode[KeyCode.F5] = 'F5';
    KeyCode[KeyCode.F6] = 'F6';
    KeyCode[KeyCode.F7] = 'F7';
    KeyCode[KeyCode.F8] = 'F8';
    KeyCode[KeyCode.F9] = 'F9';
    KeyCode[KeyCode.F10] = 'F10';
    KeyCode[KeyCode.F11] = 'F11';
    KeyCode[KeyCode.F12] = 'F12';
    KeyCode[KeyCode.NumLock] = 'NumLock';
    KeyCode[KeyCode.ScrollLock] = 'ScrollLock';
    KeyCode[KeyCode.SemiColon] = 'SemiColon';
    KeyCode[KeyCode.Equals] = 'Equals';
    KeyCode[KeyCode.Comma] = 'Comma';
    KeyCode[KeyCode.Dash] = 'Dash';
    KeyCode[KeyCode.Period] = 'Period';
    KeyCode[KeyCode.UnderScore] = 'UnderScore';
    KeyCode[KeyCode.PlusSign] = 'PlusSign';
    KeyCode[KeyCode.ForwardSlash] = 'ForwardSlash';
    KeyCode[KeyCode.Tilde] = 'Tilde';
    KeyCode[KeyCode.GraveAccent] = 'GraveAccent';
    KeyCode[KeyCode.OpenBracket] = 'OpenBracket';
    KeyCode[KeyCode.ClosedBracket] = 'ClosedBracket';
    KeyCode[KeyCode.Quote] = 'Quote';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ScrollService = (function () {
        function ScrollService(element) {
            this.element = element;
        }
        /**
         * @param {?} activeIndex
         * @return {?}
         */
        ScrollService.prototype.handleKeyboarOptionNavigation = /**
         * @param {?} activeIndex
         * @return {?}
         */
            function (activeIndex) {
                /** @type {?} */
                var autocompleteElement = ((this.element.nativeElement));
                /** @type {?} */
                var listElement = autocompleteElement.querySelector("." + SEARCH_WRAPPER_CLASS);
                if (listElement) {
                    if (activeIndex === DEFAULT_ACTIVE_INDEX) {
                        listElement.scrollTo(0, 0);
                        return;
                    }
                    /** @type {?} */
                    var optionElements = listElement.querySelectorAll("." + OPTION_CLASS);
                    /** @type {?} */
                    var activeElement = optionElements[activeIndex];
                    /** @type {?} */
                    var listTop = listElement.getBoundingClientRect().top;
                    /** @type {?} */
                    var listBottom = listTop + listElement.clientHeight;
                    /** @type {?} */
                    var elementTop = activeElement.getBoundingClientRect().top;
                    /** @type {?} */
                    var elementBottom = elementTop + activeElement.clientHeight;
                    if (listTop > elementTop) {
                        activeElement.scrollIntoView(true);
                    }
                    if (listBottom < elementBottom) {
                        activeElement.scrollIntoView(false);
                    }
                }
            };
        return ScrollService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var KeyboardNavigationDirective = (function () {
        function KeyboardNavigationDirective(autocompleteService, itemListService, element) {
            this.autocompleteService = autocompleteService;
            this.itemListService = itemListService;
            this.element = element;
            this.stopEnterPropagation = false;
            this.scrollService = new ScrollService(element);
        }
        /**
         * @param {?} event
         * @return {?}
         */
        KeyboardNavigationDirective.prototype.onKeydown = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                /** @type {?} */
                var key = event.which || event.keyCode;
                switch (key) {
                    case KeyCode.DownArrow:
                        this.handleDownKey(event);
                        break;
                    case KeyCode.UpArrow:
                        this.handleUpKey(event);
                        break;
                    case KeyCode.Enter:
                        this.handleEnter(event);
                        break;
                }
            };
        /**
         * @private
         * @param {?} event
         * @return {?}
         */
        KeyboardNavigationDirective.prototype.handleDownKey = /**
         * @private
         * @param {?} event
         * @return {?}
         */
            function (event) {
                this.itemListService.activeIndex++;
                this.scrollService.handleKeyboarOptionNavigation(this.itemListService.activeIndex);
            };
        /**
         * @private
         * @param {?} event
         * @return {?}
         */
        KeyboardNavigationDirective.prototype.handleUpKey = /**
         * @private
         * @param {?} event
         * @return {?}
         */
            function (event) {
                this.itemListService.activeIndex--;
                this.scrollService.handleKeyboarOptionNavigation(this.itemListService.activeIndex);
            };
        /**
         * @private
         * @param {?} event
         * @return {?}
         */
        KeyboardNavigationDirective.prototype.handleEnter = /**
         * @private
         * @param {?} event
         * @return {?}
         */
            function (event) {
                if (this.itemListService.activeIndex >= 0) {
                    this.autocompleteService.selectItemSubject.next(this.itemListService.activeIndex);
                    event.preventDefault();
                    if (this.stopEnterPropagation) {
                        event.stopPropagation();
                    }
                }
            };
        KeyboardNavigationDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[keyboard-navigation]'
                    },] },
        ];
        KeyboardNavigationDirective.ctorParameters = function () {
            return [
                { type: AutocompleteService },
                { type: ItemListService },
                { type: core.ElementRef }
            ];
        };
        KeyboardNavigationDirective.propDecorators = {
            stopEnterPropagation: [{ type: core.Input }],
            onKeydown: [{ type: core.HostListener, args: ['keydown', ['$event'],] }]
        };
        return KeyboardNavigationDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var RemoteAutocompleteModule = (function () {
        function RemoteAutocompleteModule() {
        }
        RemoteAutocompleteModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [
                            common.CommonModule,
                            forms.FormsModule
                        ],
                        declarations: [
                            RemoteAutocompleteComponent,
                            OptionTemplateDirective,
                            OptionDirective,
                            KeyboardNavigationDirective
                        ],
                        exports: [
                            RemoteAutocompleteComponent,
                            OptionTemplateDirective
                        ],
                    },] },
        ];
        return RemoteAutocompleteModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.RemoteAutocompleteModule = RemoteAutocompleteModule;
    exports.RemoteAutocompleteComponent = RemoteAutocompleteComponent;
    exports.OptionTemplateDirective = OptionTemplateDirective;
    exports.ɵd = KeyboardNavigationDirective;
    exports.ɵc = OptionDirective;
    exports.ɵa = AutocompleteService;
    exports.ɵb = ItemListService;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
