// ng-remote-autocomplete/esm5/app/directives/keyboard-navigation.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostListener, ElementRef, Input } from "@angular/core";
import { KeyCode } from "../enums/key-kode";
import { ItemListService } from "../services/item-list.service";
import { AutocompleteService } from "../services/autocomplete.service";
import { ScrollService } from "../services/scroll-service";
var KeyboardNavigationDirective = /** @class */ (function () {
    function KeyboardNavigationDirective(autocompleteService, itemListService, element) {
        this.autocompleteService = autocompleteService;
        this.itemListService = itemListService;
        this.element = element;
        this.stopEnterPropagation = false;
        this.scrollService = new ScrollService(element);
    }
    /**
     * @param {?} event
     * @return {?}
     */
    KeyboardNavigationDirective.prototype.onKeydown = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var key = event.which || event.keyCode;
        switch (key) {
            case KeyCode.DownArrow:
                this.handleDownKey(event);
                break;
            case KeyCode.UpArrow:
                this.handleUpKey(event);
                break;
            case KeyCode.Enter:
                this.handleEnter(event);
                break;
        }
    };
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    KeyboardNavigationDirective.prototype.handleDownKey = /**
     * @private
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.itemListService.activeIndex++;
        this.scrollService.handleKeyboarOptionNavigation(this.itemListService.activeIndex);
    };
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    KeyboardNavigationDirective.prototype.handleUpKey = /**
     * @private
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.itemListService.activeIndex--;
        this.scrollService.handleKeyboarOptionNavigation(this.itemListService.activeIndex);
    };
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    KeyboardNavigationDirective.prototype.handleEnter = /**
     * @private
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.itemListService.activeIndex >= 0) {
            this.autocompleteService.selectItemSubject.next(this.itemListService.activeIndex);
            event.preventDefault();
            if (this.stopEnterPropagation) {
                event.stopPropagation();
            }
        }
    };
    KeyboardNavigationDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[keyboard-navigation]'
                },] },
    ];
    KeyboardNavigationDirective.ctorParameters = function () { return [
        { type: AutocompleteService },
        { type: ItemListService },
        { type: ElementRef }
    ]; };
    KeyboardNavigationDirective.propDecorators = {
        stopEnterPropagation: [{ type: Input }],
        onKeydown: [{ type: HostListener, args: ['keydown', ['$event'],] }]
    };
    return KeyboardNavigationDirective;
}());
export { KeyboardNavigationDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    KeyboardNavigationDirective.prototype.scrollService;
    /** @type {?} */
    KeyboardNavigationDirective.prototype.stopEnterPropagation;
    /**
     * @type {?}
     * @private
     */
    KeyboardNavigationDirective.prototype.autocompleteService;
    /**
     * @type {?}
     * @private
     */
    KeyboardNavigationDirective.prototype.itemListService;
    /**
     * @type {?}
     * @private
     */
    KeyboardNavigationDirective.prototype.element;
}
