// ng-remote-autocomplete/esm5/app/directives/option-directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input, Directive } from "@angular/core";
import { ItemListService } from "../services/item-list.service";
import { AutocompleteService } from "../services/autocomplete.service";
var OptionDirective = /** @class */ (function () {
    function OptionDirective(autocompleteService, itemListService) {
        this.autocompleteService = autocompleteService;
        this.itemListService = itemListService;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    OptionDirective.prototype.onMouseEnter = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.itemListService.activeIndex = this.index;
    };
    /**
     * @param {?} event
     * @return {?}
     */
    OptionDirective.prototype.onMouseLeave = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.itemListService.restoreIntialActiveIndex();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    OptionDirective.prototype.onClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.autocompleteService.selectItemSubject.next(this.index);
    };
    OptionDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[remote-autocomplete-option]',
                    host: {
                        '(mouseenter)': 'onMouseEnter($event)',
                        '(mouseleave)': 'onMouseLeave($event)',
                        '(click)': 'onClick($event)'
                    }
                },] },
    ];
    OptionDirective.ctorParameters = function () { return [
        { type: AutocompleteService },
        { type: ItemListService }
    ]; };
    OptionDirective.propDecorators = {
        index: [{ type: Input, args: ['itemIndex',] }]
    };
    return OptionDirective;
}());
export { OptionDirective };
if (false) {
    /** @type {?} */
    OptionDirective.prototype.index;
    /**
     * @type {?}
     * @private
     */
    OptionDirective.prototype.autocompleteService;
    /**
     * @type {?}
     * @private
     */
    OptionDirective.prototype.itemListService;
}
