// ng-remote-autocomplete/esm5/app/constants/autocomplete.constants.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var DEFAULT_ACTIVE_INDEX = -1;
/** @type {?} */
export var DEFAULT_MIN_SEARCH_LENGTH = 1;
/** @type {?} */
export var DEFAULT_MAX_CHARS = 2147483647;
/** @type {?} */
export var DEFAULT_PAUSE = 100;
/** @type {?} */
export var NOT_FOUND_TEXT = "No results found";
/** @type {?} */
export var SEARCHING_TEXT = "Searching...";
/** @type {?} */
export var SEARCH_WRAPPER_CLASS = "autocomplete-search-wrapper";
/** @type {?} */
export var OPTION_VALUE_CLASS = "autocomplete-option-value";
/** @type {?} */
export var OPTION_DESCRIPTION_CLASS = "autocomplete-option-description";
/** @type {?} */
export var OPTION_CLASS = "autocomplete-option";
/** @type {?} */
export var OPTION_LIST_WRAPPER_CLASS = "option-list-wrapper";
/** @type {?} */
export var LOADING_CLASS = "autocomplete-loading";
/** @type {?} */
export var RESULTS_NOT_FOUND_CLASS = "autocomplete-not-found";
