// ng-remote-autocomplete/esm5/app/components/autocomplete.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ContentChild, TemplateRef, forwardRef, ViewChild, ElementRef } from "@angular/core";
import { NG_VALUE_ACCESSOR } from "@angular/forms";
import { DEFAULT_MAX_CHARS, DEFAULT_MIN_SEARCH_LENGTH, DEFAULT_PAUSE, NOT_FOUND_TEXT, SEARCHING_TEXT, DEFAULT_ACTIVE_INDEX, SEARCH_WRAPPER_CLASS, OPTION_LIST_WRAPPER_CLASS, OPTION_CLASS, OPTION_VALUE_CLASS, OPTION_DESCRIPTION_CLASS, LOADING_CLASS, RESULTS_NOT_FOUND_CLASS } from "../constants/autocomplete.constants";
import { SearchStateType } from "../enums/search-state.type";
import { OptionTemplateDirective } from "../directives/option-template.directive";
import { ItemListService } from "../services/item-list.service";
import { AutocompleteService } from "../services/autocomplete.service";
var RemoteAutocompleteComponent = /** @class */ (function () {
    function RemoteAutocompleteComponent(itemListService, autocompleteService) {
        this.itemListService = itemListService;
        this.autocompleteService = autocompleteService;
        this.stopEnterPropagationOnSelect = false;
        this.needToShowNotFoundText = true;
        this.needToShowSearchingText = true;
        this.type = new EventEmitter();
        this.highlighted = new EventEmitter();
        this.selected = new EventEmitter();
        this.blur = new EventEmitter();
        this.focusEvent = new EventEmitter();
        this.searchStates = SearchStateType;
        this.searchState = SearchStateType.Untracked;
        this.SEARCH_WRAPPER_CLASS = SEARCH_WRAPPER_CLASS;
        this.OPTION_LIST_WRAPPER_CLASS = OPTION_LIST_WRAPPER_CLASS;
        this.OPTION_CLASS = OPTION_CLASS;
        this.OPTION_VALUE_CLASS = OPTION_VALUE_CLASS;
        this.OPTION_DESCRIPTION_CLASS = OPTION_DESCRIPTION_CLASS;
        this.LOADING_CLASS = LOADING_CLASS;
        this.RESULTS_NOT_FOUND_CLASS = RESULTS_NOT_FOUND_CLASS;
        this.propagateChange = (/**
         * @param {?} _
         * @return {?}
         */
        function (_) { });
        this.propagateTouched = (/**
         * @param {?} _
         * @return {?}
         */
        function (_) { });
    }
    /**
     * @return {?}
     */
    RemoteAutocompleteComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.validateService();
        this.disabled = this.getDefaultIfNullOrUndefined(this.disabled, false);
        this.minChars = this.getDefaultIfNullOrUndefined(this.minChars, DEFAULT_MIN_SEARCH_LENGTH);
        this.maxChars = this.maxChars || DEFAULT_MAX_CHARS;
        this.pause = this.pause || DEFAULT_PAUSE;
        this.notFoundText = this.notFoundText || NOT_FOUND_TEXT;
        this.searchingText = this.searchingText || SEARCHING_TEXT;
        this.changedHighlightSubscription = this.autocompleteService.changedHighlightItemSubject.subscribe((/**
         * @param {?} index
         * @return {?}
         */
        function (index) {
            _this.onActiveIndexChanged(index);
        }));
        this.selectSubscription = this.autocompleteService.selectItemSubject.subscribe((/**
         * @param {?} index
         * @return {?}
         */
        function (index) {
            /** @type {?} */
            var result = _this.searchResult[index];
            _this.propagateChange(result.value);
            _this.selected.emit(result);
            _this.close();
        }));
    };
    /**
     * @return {?}
     */
    RemoteAutocompleteComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.changedHighlightSubscription.unsubscribe();
        this.selectSubscription.unsubscribe();
    };
    /**
     * @param {?} value
     * @return {?}
     */
    RemoteAutocompleteComponent.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value !== undefined) {
            this.searchValue = value;
        }
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    RemoteAutocompleteComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.propagateChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    RemoteAutocompleteComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.propagateTouched = fn;
    };
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    RemoteAutocompleteComponent.prototype.setDisabledState = /**
     * @param {?} isDisabled
     * @return {?}
     */
    function (isDisabled) {
        this.disabled = isDisabled;
    };
    /**
     * @return {?}
     */
    RemoteAutocompleteComponent.prototype.onType = /**
     * @return {?}
     */
    function () {
        this.propagateChange(this.searchValue);
        this.originalSearchValue = this.searchValue;
        this.type.emit();
        if (this.validsearchTerm()) {
            this.searchState = this.searchStates.Loading;
        }
        else {
            this.close();
        }
        this.resetSearchTimeout();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    RemoteAutocompleteComponent.prototype.onInputBlur = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (event.relatedTarget) {
            /** @type {?} */
            var related = (/** @type {?} */ (event.relatedTarget));
            if (related.classList.contains(this.OPTION_CLASS)) {
                return;
            }
        }
        this.propagateTouched(this.searchValue);
        this.close();
        this.blur.emit(event);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    RemoteAutocompleteComponent.prototype.onInputFocus = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.focusEvent.emit(event);
    };
    Object.defineProperty(RemoteAutocompleteComponent.prototype, "inputClass", {
        get: /**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var classes = '';
            if (this.inputClasses) {
                classes = this.inputClasses.join(' ');
            }
            return classes;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    RemoteAutocompleteComponent.prototype.focus = /**
     * @return {?}
     */
    function () {
        ((/** @type {?} */ (this.inputField.nativeElement))).focus();
    };
    /**
     * @return {?}
     */
    RemoteAutocompleteComponent.prototype.open = /**
     * @return {?}
     */
    function () {
        this.onType();
    };
    /**
     * @return {?}
     */
    RemoteAutocompleteComponent.prototype.close = /**
     * @return {?}
     */
    function () {
        this.autocompleteService.isOpen = false;
        this.searchState = this.searchStates.Untracked;
    };
    /**
     * @private
     * @return {?}
     */
    RemoteAutocompleteComponent.prototype.validateService = /**
     * @private
     * @return {?}
     */
    function () {
        if (!this.service) {
            throw new Error("AutocompleteService must be implemented");
        }
        if (this.isParameterizedSearch(this.service) && (this.searchParameters === undefined || this.searchParameters === null)) {
            throw new Error("Please provide additionalSearchParams or just provide AutocompleteSourceService instaead of ParameterizedAutocompleteSourceService");
        }
    };
    /**
     * @private
     * @param {?} activeIndex
     * @return {?}
     */
    RemoteAutocompleteComponent.prototype.onActiveIndexChanged = /**
     * @private
     * @param {?} activeIndex
     * @return {?}
     */
    function (activeIndex) {
        /** @type {?} */
        var highlightedItem;
        if (activeIndex === DEFAULT_ACTIVE_INDEX) {
            highlightedItem = null;
            this.searchValue = this.originalSearchValue;
        }
        else {
            highlightedItem = this.searchResult[activeIndex];
            this.searchValue = highlightedItem.value;
        }
        this.propagateChange(this.searchValue);
        this.highlighted.emit(highlightedItem);
    };
    /**
     * @private
     * @return {?}
     */
    RemoteAutocompleteComponent.prototype.resetSearchTimeout = /**
     * @private
     * @return {?}
     */
    function () {
        this.clearSearchTimeout();
        this.setSearchTimeout();
    };
    /**
     * @private
     * @return {?}
     */
    RemoteAutocompleteComponent.prototype.setSearchTimeout = /**
     * @private
     * @return {?}
     */
    function () {
        this.typingTimeout = setTimeout(this.search.bind(this), this.pause);
    };
    /**
     * @private
     * @return {?}
     */
    RemoteAutocompleteComponent.prototype.clearSearchTimeout = /**
     * @private
     * @return {?}
     */
    function () {
        clearTimeout(this.typingTimeout);
    };
    /**
     * @private
     * @return {?}
     */
    RemoteAutocompleteComponent.prototype.search = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.itemListService.restoreIntialActiveIndex();
        if (this.validsearchTerm()) {
            this.performSearch().first().subscribe((/**
             * @param {?} results
             * @return {?}
             */
            function (results) {
                if (_this.searchState !== _this.searchStates.Untracked) {
                    _this.autocompleteService.isOpen = results.length > 0;
                    _this.itemListService.items = results;
                    _this.searchResult = results;
                    _this.searchState = _this.searchStates.Finished;
                }
            }));
        }
    };
    /**
     * @private
     * @return {?}
     */
    RemoteAutocompleteComponent.prototype.performSearch = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.isParameterizedSearch(this.service)) {
            return ((/** @type {?} */ (this.service))).getWithParams(this.searchValue, this.searchParameters);
        }
        else {
            return ((/** @type {?} */ (this.service))).get(this.searchValue);
        }
    };
    /**
     * @private
     * @return {?}
     */
    RemoteAutocompleteComponent.prototype.validsearchTerm = /**
     * @private
     * @return {?}
     */
    function () {
        return (this.searchValue || '').length >= this.minChars;
    };
    /**
     * @private
     * @param {?} service
     * @return {?}
     */
    RemoteAutocompleteComponent.prototype.isParameterizedSearch = /**
     * @private
     * @param {?} service
     * @return {?}
     */
    function (service) {
        return ((/** @type {?} */ (service))).getWithParams !== undefined;
    };
    /**
     * @private
     * @template T
     * @param {?} value
     * @param {?} defaultValue
     * @return {?}
     */
    RemoteAutocompleteComponent.prototype.getDefaultIfNullOrUndefined = /**
     * @private
     * @template T
     * @param {?} value
     * @param {?} defaultValue
     * @return {?}
     */
    function (value, defaultValue) {
        return value === undefined || value === null
            ? defaultValue
            : value;
    };
    RemoteAutocompleteComponent.decorators = [
        { type: Component, args: [{
                    selector: 'remote-autocomplete',
                    template: "<div keyboard-navigation [stopEnterPropagation]=\"stopEnterPropagationOnSelect\" class=\"remote-autocomplete\">\n    <input [(ngModel)]=\"searchValue\"\n           [attr.id]=\"inputId\"\n           [attr.name]=\"inputName\"\n           [attr.class]=\"inputClass\"\n           [disabled]=\"disabled\"\n           [attr.placeholder]=\"placeholder\"\n           [attr.autocomplete]=\"autocompleteAttr\"\n           [attr.tabindex]=\"tabindex\"\n           maxlength=\"{{maxChars}}\"\n           (input)=\"onType()\"\n           (blur)=\"onInputBlur($event)\"\n           (focus)=\"onInputFocus($event)\"\n           #input />\n    <div *ngIf=\"searchState === searchStates.Loading && needToShowSearchingText || \n                searchState === searchStates.Finished && (searchResult?.length > 0 || needToShowNotFoundText)\" [class]=\"SEARCH_WRAPPER_CLASS\">\n        <div *ngIf=\"searchState === searchStates.Loading\">\n            <span [class]=\"LOADING_CLASS\">{{searchingText}}</span>\n        </div>\n\n        <div *ngIf=\"searchState === searchStates.Finished && searchResult.length === 0\">\n            <span [class]=\"RESULTS_NOT_FOUND_CLASS\">{{notFoundText}}</span>\n        </div>\n\n        <div *ngIf=\"searchState === searchStates.Finished && searchResult?.length > 0\" [class]=\"OPTION_LIST_WRAPPER_CLASS\">\n            <div *ngFor=\"let item of searchResult; index as i\"\n                 remote-autocomplete-option\n                 [itemIndex]=\"i\"\n                 [class]=\"OPTION_CLASS\"\n                 [ngClass]=\"{'autocomplete-option-active': i === itemListService.activeIndex}\"\n                 tabindex=\"-1\">\n                <ng-template #defaultOptionTemplate>\n                    <span [class]=\"OPTION_VALUE_CLASS\">{{item.value}}</span>\n                    <span *ngIf=\"item.description\" [class]=\"OPTION_DESCRIPTION_CLASS\">{{item.description}}</span>\n                </ng-template>\n\n                <ng-template [ngTemplateOutlet]=\"optionTemplate || defaultOptionTemplate\" [ngTemplateOutletContext]=\"{ $implicit: item }\">\n                </ng-template>\n            </div>\n        </div>\n    </div>\n</div>",
                    styles: [".autocomplete-search-wrapper{border:1px solid #ccc;max-height:200px;overflow:auto;max-width:200px}.autocomplete-option-active{background-color:#d3d3d3}"],
                    providers: [
                        AutocompleteService,
                        ItemListService,
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return RemoteAutocompleteComponent; })),
                            multi: true
                        }
                    ]
                },] },
    ];
    RemoteAutocompleteComponent.ctorParameters = function () { return [
        { type: ItemListService },
        { type: AutocompleteService }
    ]; };
    RemoteAutocompleteComponent.propDecorators = {
        service: [{ type: Input, args: ['service',] }],
        minChars: [{ type: Input, args: ['minSearchLength',] }],
        maxChars: [{ type: Input, args: ['maxChars',] }],
        pause: [{ type: Input, args: ['pause',] }],
        inputId: [{ type: Input, args: ['inputId',] }],
        inputName: [{ type: Input, args: ['inputName',] }],
        inputClasses: [{ type: Input, args: ['inputClass',] }],
        disabled: [{ type: Input, args: ['disabled',] }],
        notFoundText: [{ type: Input, args: ['notFoundText',] }],
        searchingText: [{ type: Input, args: ['searchingText',] }],
        placeholder: [{ type: Input, args: ['placeholder',] }],
        searchParameters: [{ type: Input, args: ['searchParameters',] }],
        autocompleteAttr: [{ type: Input, args: ['autocompleteAttr',] }],
        tabindex: [{ type: Input, args: ['tabindex',] }],
        stopEnterPropagationOnSelect: [{ type: Input, args: ['stopEnterPropagationOnSelect',] }],
        needToShowNotFoundText: [{ type: Input, args: ['needToShowNotFoundText',] }],
        needToShowSearchingText: [{ type: Input, args: ['needToShowSearchingText',] }],
        type: [{ type: Output, args: ['type',] }],
        highlighted: [{ type: Output, args: ['highlighted',] }],
        selected: [{ type: Output, args: ['selected',] }],
        blur: [{ type: Output, args: ['blur',] }],
        focusEvent: [{ type: Output, args: ['focus',] }],
        optionTemplate: [{ type: ContentChild, args: [OptionTemplateDirective, { read: TemplateRef },] }],
        inputField: [{ type: ViewChild, args: ['input',] }]
    };
    return RemoteAutocompleteComponent;
}());
export { RemoteAutocompleteComponent };
if (false) {
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.service;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.minChars;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.maxChars;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.pause;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.inputId;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.inputName;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.inputClasses;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.disabled;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.notFoundText;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.searchingText;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.placeholder;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.searchParameters;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.autocompleteAttr;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.tabindex;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.stopEnterPropagationOnSelect;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.needToShowNotFoundText;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.needToShowSearchingText;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.type;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.highlighted;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.selected;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.blur;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.focusEvent;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.optionTemplate;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.inputField;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.searchStates;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.searchState;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.searchValue;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.searchResult;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.SEARCH_WRAPPER_CLASS;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.OPTION_LIST_WRAPPER_CLASS;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.OPTION_CLASS;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.OPTION_VALUE_CLASS;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.OPTION_DESCRIPTION_CLASS;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.LOADING_CLASS;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.RESULTS_NOT_FOUND_CLASS;
    /**
     * @type {?}
     * @private
     */
    RemoteAutocompleteComponent.prototype.typingTimeout;
    /**
     * @type {?}
     * @private
     */
    RemoteAutocompleteComponent.prototype.originalSearchValue;
    /**
     * @type {?}
     * @private
     */
    RemoteAutocompleteComponent.prototype.changedHighlightSubscription;
    /**
     * @type {?}
     * @private
     */
    RemoteAutocompleteComponent.prototype.selectSubscription;
    /**
     * @type {?}
     * @private
     */
    RemoteAutocompleteComponent.prototype.propagateChange;
    /**
     * @type {?}
     * @private
     */
    RemoteAutocompleteComponent.prototype.propagateTouched;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.itemListService;
    /** @type {?} */
    RemoteAutocompleteComponent.prototype.autocompleteService;
}
