// ng-remote-autocomplete/esm2015/app/enums/key-kode.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
const KeyCode = {
    Backspace: 8,
    Tab: 9,
    Enter: 13,
    Shift: 16,
    Ctrl: 17,
    Alt: 18,
    PauseBreak: 19,
    CapsLock: 20,
    Escape: 27,
    Space: 32,
    PageUp: 33,
    PageDown: 34,
    End: 35,
    Home: 36,
    LeftArrow: 37,
    UpArrow: 38,
    RightArrow: 39,
    DownArrow: 40,
    Insert: 45,
    Delete: 46,
    Zero: 48,
    ClosedParen: 48,
    One: 49,
    ExclamationMark: 49,
    Two: 50,
    AtSign: 50,
    Three: 51,
    PoundSign: 51,
    Hash: 51,
    Four: 52,
    DollarSign: 52,
    Five: 53,
    PercentSign: 53,
    Six: 54,
    Caret: 54,
    Hat: 54,
    Seven: 55,
    Ampersand: 55,
    Eight: 56,
    Star: 56,
    Asterik: 56,
    Nine: 57,
    OpenParen: 57,
    A: 65,
    B: 66,
    C: 67,
    D: 68,
    E: 69,
    F: 70,
    G: 71,
    H: 72,
    I: 73,
    J: 74,
    K: 75,
    L: 76,
    M: 77,
    N: 78,
    O: 79,
    P: 80,
    Q: 81,
    R: 82,
    S: 83,
    T: 84,
    U: 85,
    V: 86,
    W: 87,
    X: 88,
    Y: 89,
    Z: 90,
    LeftWindowKey: 91,
    RightWindowKey: 92,
    SelectKey: 93,
    Numpad0: 96,
    Numpad1: 97,
    Numpad2: 98,
    Numpad3: 99,
    Numpad4: 100,
    Numpad5: 101,
    Numpad6: 102,
    Numpad7: 103,
    Numpad8: 104,
    Numpad9: 105,
    Multiply: 106,
    Add: 107,
    Subtract: 109,
    DecimalPoint: 110,
    Divide: 111,
    F1: 112,
    F2: 113,
    F3: 114,
    F4: 115,
    F5: 116,
    F6: 117,
    F7: 118,
    F8: 119,
    F9: 120,
    F10: 121,
    F11: 122,
    F12: 123,
    NumLock: 144,
    ScrollLock: 145,
    SemiColon: 186,
    Equals: 187,
    Comma: 188,
    Dash: 189,
    Period: 190,
    UnderScore: 189,
    PlusSign: 187,
    ForwardSlash: 191,
    Tilde: 192,
    GraveAccent: 192,
    OpenBracket: 219,
    ClosedBracket: 221,
    Quote: 222,
};
export { KeyCode };
KeyCode[KeyCode.Backspace] = 'Backspace';
KeyCode[KeyCode.Tab] = 'Tab';
KeyCode[KeyCode.Enter] = 'Enter';
KeyCode[KeyCode.Shift] = 'Shift';
KeyCode[KeyCode.Ctrl] = 'Ctrl';
KeyCode[KeyCode.Alt] = 'Alt';
KeyCode[KeyCode.PauseBreak] = 'PauseBreak';
KeyCode[KeyCode.CapsLock] = 'CapsLock';
KeyCode[KeyCode.Escape] = 'Escape';
KeyCode[KeyCode.Space] = 'Space';
KeyCode[KeyCode.PageUp] = 'PageUp';
KeyCode[KeyCode.PageDown] = 'PageDown';
KeyCode[KeyCode.End] = 'End';
KeyCode[KeyCode.Home] = 'Home';
KeyCode[KeyCode.LeftArrow] = 'LeftArrow';
KeyCode[KeyCode.UpArrow] = 'UpArrow';
KeyCode[KeyCode.RightArrow] = 'RightArrow';
KeyCode[KeyCode.DownArrow] = 'DownArrow';
KeyCode[KeyCode.Insert] = 'Insert';
KeyCode[KeyCode.Delete] = 'Delete';
KeyCode[KeyCode.Zero] = 'Zero';
KeyCode[KeyCode.ClosedParen] = 'ClosedParen';
KeyCode[KeyCode.One] = 'One';
KeyCode[KeyCode.ExclamationMark] = 'ExclamationMark';
KeyCode[KeyCode.Two] = 'Two';
KeyCode[KeyCode.AtSign] = 'AtSign';
KeyCode[KeyCode.Three] = 'Three';
KeyCode[KeyCode.PoundSign] = 'PoundSign';
KeyCode[KeyCode.Hash] = 'Hash';
KeyCode[KeyCode.Four] = 'Four';
KeyCode[KeyCode.DollarSign] = 'DollarSign';
KeyCode[KeyCode.Five] = 'Five';
KeyCode[KeyCode.PercentSign] = 'PercentSign';
KeyCode[KeyCode.Six] = 'Six';
KeyCode[KeyCode.Caret] = 'Caret';
KeyCode[KeyCode.Hat] = 'Hat';
KeyCode[KeyCode.Seven] = 'Seven';
KeyCode[KeyCode.Ampersand] = 'Ampersand';
KeyCode[KeyCode.Eight] = 'Eight';
KeyCode[KeyCode.Star] = 'Star';
KeyCode[KeyCode.Asterik] = 'Asterik';
KeyCode[KeyCode.Nine] = 'Nine';
KeyCode[KeyCode.OpenParen] = 'OpenParen';
KeyCode[KeyCode.A] = 'A';
KeyCode[KeyCode.B] = 'B';
KeyCode[KeyCode.C] = 'C';
KeyCode[KeyCode.D] = 'D';
KeyCode[KeyCode.E] = 'E';
KeyCode[KeyCode.F] = 'F';
KeyCode[KeyCode.G] = 'G';
KeyCode[KeyCode.H] = 'H';
KeyCode[KeyCode.I] = 'I';
KeyCode[KeyCode.J] = 'J';
KeyCode[KeyCode.K] = 'K';
KeyCode[KeyCode.L] = 'L';
KeyCode[KeyCode.M] = 'M';
KeyCode[KeyCode.N] = 'N';
KeyCode[KeyCode.O] = 'O';
KeyCode[KeyCode.P] = 'P';
KeyCode[KeyCode.Q] = 'Q';
KeyCode[KeyCode.R] = 'R';
KeyCode[KeyCode.S] = 'S';
KeyCode[KeyCode.T] = 'T';
KeyCode[KeyCode.U] = 'U';
KeyCode[KeyCode.V] = 'V';
KeyCode[KeyCode.W] = 'W';
KeyCode[KeyCode.X] = 'X';
KeyCode[KeyCode.Y] = 'Y';
KeyCode[KeyCode.Z] = 'Z';
KeyCode[KeyCode.LeftWindowKey] = 'LeftWindowKey';
KeyCode[KeyCode.RightWindowKey] = 'RightWindowKey';
KeyCode[KeyCode.SelectKey] = 'SelectKey';
KeyCode[KeyCode.Numpad0] = 'Numpad0';
KeyCode[KeyCode.Numpad1] = 'Numpad1';
KeyCode[KeyCode.Numpad2] = 'Numpad2';
KeyCode[KeyCode.Numpad3] = 'Numpad3';
KeyCode[KeyCode.Numpad4] = 'Numpad4';
KeyCode[KeyCode.Numpad5] = 'Numpad5';
KeyCode[KeyCode.Numpad6] = 'Numpad6';
KeyCode[KeyCode.Numpad7] = 'Numpad7';
KeyCode[KeyCode.Numpad8] = 'Numpad8';
KeyCode[KeyCode.Numpad9] = 'Numpad9';
KeyCode[KeyCode.Multiply] = 'Multiply';
KeyCode[KeyCode.Add] = 'Add';
KeyCode[KeyCode.Subtract] = 'Subtract';
KeyCode[KeyCode.DecimalPoint] = 'DecimalPoint';
KeyCode[KeyCode.Divide] = 'Divide';
KeyCode[KeyCode.F1] = 'F1';
KeyCode[KeyCode.F2] = 'F2';
KeyCode[KeyCode.F3] = 'F3';
KeyCode[KeyCode.F4] = 'F4';
KeyCode[KeyCode.F5] = 'F5';
KeyCode[KeyCode.F6] = 'F6';
KeyCode[KeyCode.F7] = 'F7';
KeyCode[KeyCode.F8] = 'F8';
KeyCode[KeyCode.F9] = 'F9';
KeyCode[KeyCode.F10] = 'F10';
KeyCode[KeyCode.F11] = 'F11';
KeyCode[KeyCode.F12] = 'F12';
KeyCode[KeyCode.NumLock] = 'NumLock';
KeyCode[KeyCode.ScrollLock] = 'ScrollLock';
KeyCode[KeyCode.SemiColon] = 'SemiColon';
KeyCode[KeyCode.Equals] = 'Equals';
KeyCode[KeyCode.Comma] = 'Comma';
KeyCode[KeyCode.Dash] = 'Dash';
KeyCode[KeyCode.Period] = 'Period';
KeyCode[KeyCode.UnderScore] = 'UnderScore';
KeyCode[KeyCode.PlusSign] = 'PlusSign';
KeyCode[KeyCode.ForwardSlash] = 'ForwardSlash';
KeyCode[KeyCode.Tilde] = 'Tilde';
KeyCode[KeyCode.GraveAccent] = 'GraveAccent';
KeyCode[KeyCode.OpenBracket] = 'OpenBracket';
KeyCode[KeyCode.ClosedBracket] = 'ClosedBracket';
KeyCode[KeyCode.Quote] = 'Quote';
