// ng-remote-autocomplete/esm2015/app/constants/autocomplete.constants.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const DEFAULT_ACTIVE_INDEX = -1;
/** @type {?} */
export const DEFAULT_MIN_SEARCH_LENGTH = 1;
/** @type {?} */
export const DEFAULT_MAX_CHARS = 2147483647;
/** @type {?} */
export const DEFAULT_PAUSE = 100;
/** @type {?} */
export const NOT_FOUND_TEXT = "No results found";
/** @type {?} */
export const SEARCHING_TEXT = "Searching...";
/** @type {?} */
export const SEARCH_WRAPPER_CLASS = "autocomplete-search-wrapper";
/** @type {?} */
export const OPTION_VALUE_CLASS = "autocomplete-option-value";
/** @type {?} */
export const OPTION_DESCRIPTION_CLASS = "autocomplete-option-description";
/** @type {?} */
export const OPTION_CLASS = "autocomplete-option";
/** @type {?} */
export const OPTION_LIST_WRAPPER_CLASS = "option-list-wrapper";
/** @type {?} */
export const LOADING_CLASS = "autocomplete-loading";
/** @type {?} */
export const RESULTS_NOT_FOUND_CLASS = "autocomplete-not-found";
